File: pangath_properties/events/maintenance_visit.py
import frappe
from erpnext.maintenance.doctype.maintenance_visit.maintenance_visit import MaintenanceVisit

@frappe.whitelist()
def create_material_request(doc):
    main_vis=frappe.get_doc("Maintenance Visit",doc)
    met_doc=frappe.new_doc("Material Request")
    met_doc.company=main_vis.company
    met_doc.custom_maintenance_visit=main_vis.name
    for i in main_vis.purposes:
        met_doc.append("items", {
			"item_code":i.item_code,})
    return met_doc

@frappe.whitelist()
def create_sales_invoice(doc):
    main_vis=frappe.get_doc("Maintenance Visit",doc)
    sale_doc=frappe.new_doc("Sales Invoice")
    sale_doc.company=main_vis.company
    sale_doc.customer=main_vis.customer
    sale_doc.property=main_vis.custom_property
    sale_doc.unit=main_vis.custom_unit
    sale_doc.custom_maintenance_visit=main_vis.name
    for i in main_vis.purposes:
        sale_doc.append("items",{
            "item_code":i.item_code
        }
        )
    return sale_doc



@frappe.whitelist()
def create_material_issue(doc):
    main_vis=frappe.get_doc("Maintenance Visit",doc)
    sale_doc=frappe.new_doc("Stock Entry")
    sale_doc.company=main_vis.company
    sale_doc.customer=main_vis.customer
    sale_doc.stock_entry_type = "Material Issue"
    sale_doc.custom_maintenance_visit = main_vis.name
    for i in main_vis.purposes:
        sale_doc.append("items",{
            "item_code":i.item_code
        }
        )
    return sale_doc





class CustomMaintenanceVisit(MaintenanceVisit):
    def validate_purpose_table(self):
        pass


def change_status(doc,method):
    if doc.maintenance_schedule:
        maintenance_schedule = frappe.get_doc("Maintenance Schedule",doc.maintenance_schedule)
        frappe.db.set_value("Issue",maintenance_schedule.issue,"status","Resolved")

File: pangath_properties/public/js/maintenance_visit.js
frappe.ui.form.on("Maintenance Visit",{
    
    
    refresh:function(frm){
        if(frm.doc.maintenance_schedule){
            frm.set_df_property('customer', 'read_only', 1);
            frm.set_df_property('custom_tenancy_contract', 'read_only', 1);
            frm.set_df_property('custom_property', 'read_only', 1);
        }
        
        if(frm.doc.docstatus == 1){
        frm.add_custom_button(__("Material Request"), function(){
			frappe.call({
				method: 'pangath_properties.events.maintenance_visit.create_material_request',
				args: { 'doc': frm.doc.name},
				callback: function(r) {
					if (r && r.message) {pangath_properties
						var doc = frappe.model.sync(r.message);
					frappe.set_route("Form", r.message.doctype, r.message.name);
					}
				}
				
			});
		  },);
        }
        if(frm.doc.docstatus == 1){
          frm.add_custom_button(__("Sales Invoice"), function(){
			frappe.call({
				method: 'pangath_properties.events.maintenance_visit.create_sales_invoice',
				args: { 'doc': frm.doc.name },
				callback: function(r) {
					if (r && r.message) {
						var doc = frappe.model.sync(r.message);
					frappe.set_route("Form", r.message.doctype, r.message.name);
					}
				}
				
			});
		  },);
        }  
        if(frm.doc.docstatus == 1){
            frm.add_custom_button(__("Material Issue"), function(){
              frappe.call({
                  method: 'pangath_properties.events.maintenance_visit.create_material_issue',
                  args: { 'doc': frm.doc.name },
                  callback: function(r) {
                      if (r && r.message) {
                          var doc = frappe.model.sync(r.message);
                      frappe.set_route("Form", r.message.doctype, r.message.name);
                      }
                  }
                  
              });
            },);
        }
        if(frm.doc.maintenance_schedule && frm.doc.__islocal && frm.doc.issue){
            frappe.db.get_value("Maintenance Schedule",frm.doc.maintenance_schedule,"issue").then(r => {
                frappe.db.get_doc("Issue",r.message.issue).then( res =>{
                    frm.set_value("customer",res.customer)
                    frm.set_value("custom_tenancy_contract",res.custom_tenancy_contract)
                    frm.set_value("custom_property",res.property)
                    frm.set_value("custom_unit_no",res.unit)

                })
            })
        } 
    },
    
})
